### PR TITLE
perf(deepseek-v4): route ratio-128 prefill through native attention

### PR DESCRIPTION
**Pull request by DiscoStew6082**
## Summary

- route DeepSeek V4 ratio-128 compressed prefill through the existing native local-plus-pooled attention path with sequential pooled indices
- opt in only for causal/sliding masks created by the model; direct callers with custom masks retain the dense SDPA reference path
- retain dense reference attention for sub-4-bit checkpoints and preserve the existing DSpark single-token and empty-pool paths

## Performance evidence

The controlled leave-one-out ablation in #2558 measured a **10.18% marginal prompt-throughput contribution** from this path.

This is scoped to:

- Mac Studio with M3 Ultra and 512 GB unified memory
- `DeepSeek-V4-Flash-0731`, revision `7872f01b1d1fe23eabc4c98b48bffcef5a386062`
- upstream source `d2575b1df5a4012966839f75fdb200e7e0b20743`
- fixed 17,219-token prompt, SHA-256 `a1465f4b5ee68dbd173c138dd65718bd06957439b66e99069e91f50364bf81f1`
- 2,048-token prefill steps, MTP enabled with three draft tokens, cache disabled and `cached_tokens=0`

The normalized measurements and full methodology are in #2558. The linked [evidence package](https://github.com/DiscoStew6082/omlx/tree/06b515b0fd3543341b346e74cf0d3a13da2d6c06/benchmarks/evidence/deepseek_v4_flash_prefill_m3_ultra) preserves the raw investigation artifacts without adding them to this PR.

This is not an unconditional cross-chip performance claim.

## Fallbacks

- custom or otherwise unsupported masks use the original concatenate-plus-SDPA reference path
- quantization metadata below 4 bits disables this native ratio-128 route whether it appears in `quantization`, `quantization_config`, or `text_config.quantization_config`
- unsupported native shapes/dtypes, a missing native symbol, or an immediate kernel exception return directly to the original concatenate-plus-SDPA reference path without constructing the gathered sparse fallback
- DSpark single-token exact attention and empty-pool local attention are unchanged

## Validation

- `uv run pytest -q tests/test_deepseek_v4_patch.py tests/test_deepseek_v4_dspark.py tests/test_mlx_lm_mtp_patch.py tests/test_custom_kernel_abi_probe.py` — 313 passed, 1 order-dependent skip after the preceding modules initialized the base patch; the MTP file passes 119/119 alone
- native FP16 and BF16 outputs matched the causal reference within 0.004 and 0.032 maximum absolute error respectively, including 257 pooled rows across the kernel's 256-row tile boundary
- full local FP8 `DeepSeek-V4-Flash-0731` load plus a 257-token prefill and 17-token cached continuation reached the native kernel in every ratio-128 layer at both zero and nonzero offsets and produced finite logits
- full local oQ2.5 `DeepSeek-V4-Flash-0731-oQ2.5e` load plus the same cache-backed sequence produced finite logits with zero ratio-128 native-helper or kernel entries
- patched `mlx_lm.utils.load_model` tests verify the high-bit/sub-4-bit policy reaches `ModelArgs`
- all four in-tree custom-kernel ABI probes passed
- Black check passed for all changed files
- `git diff --check` passed

Refs #2558.

### CODE REVIEW

**Comment by jundot** [comment]
I reviewed and benchmarked #2559, #2562, and #2563 together since they merge cleanly as a set, so I am posting the same summary on all three. Setup: M3 Ultra 512 GB, DeepSeek-V4-Flash-0731 oQ4e (4-bit top level with mxfp4 experts, so all three changes are active), MTP off, single request, cold prefill with cached_tokens=0 verified on every run, one run per point, current main vs main plus all three PRs, using the built-in throughput benchmark (tg128).

With the prompt cache disabled, which gives 2048-token prefill steps matching the #2558 methodology:

| context | main pp tok/s | +3 PRs pp tok/s | delta |
|---:|---:|---:|---:|
| 4,096 | 464.1 | 546.3 | +17.7% |
| 16,384 | 487.8 | 559.6 | +14.7% |
| 65,536 | 463.4 | 525.1 | +13.3% |
| 131,072 | 416.1 | 460.8 | +10.7% |

tg was unchanged everywhere, which is expected for prefill-only changes. This independently reproduces the #2558 ablation on a different checkpoint, so the numbers hold up well.

One finding on my side rather than yours: with the default server config the paged SSD cache aligns the rotating block size to 512, which clamps prefill to 512-token chunks, and in that regime these gains mostly disappear (the MoE route count stays below the MXFP4 crossover and the ratio-128 win shrinks with the chunk). Raising the DeepSeek V4 block alignment to 2048 with the cache still enabled recovers the full numbers above (549.5 / 558.7 / 524.2 / 461.6 pp tok/s). I will handle that follow-up separately so default-config users actually see the improvement. Merging all three.
